lib/Dialect/AIG/Transforms/LowerCutToLUT.cpp

```diff
@@ -0,0 +1,150 @@
+//===- LowerCutToLUT.cpp ---------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This pass lowers `aig.cut` to `comb.truth_table` with k inputs where
+// k is the size of the cut (= operand inputs size).
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/AIG/AIGOps.h"
+#include "circt/Dialect/AIG/AIGPasses.h"
+#include "circt/Dialect/Comb/CombOps.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+#define DEBUG_TYPE "aig-lower-cut-to-lut"
+
+namespace circt {
+namespace aig {
+#define GEN_PASS_DEF_LOWERCUTTOLUT
+#include "circt/Dialect/AIG/AIGPasses.h.inc"
+} // namespace aig
+} // namespace circt
+
+using namespace circt;
+using namespace aig;
+
+//===----------------------------------------------------------------------===//
+// Conversion patterns
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+struct CutToLUTPattern : OpRewritePattern<CutOp> {
+  using OpRewritePattern<CutOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(CutOp op,
+                                PatternRewriter &rewriter) const override;
+};
+
+LogicalResult
+CutToLUTPattern::matchAndRewrite(CutOp cutOp, PatternRewriter &rewriter) const {
+  if (cutOp.getNumResults() == 0) {
+    rewriter.eraseOp(cutOp);
+    return success();
+  }
+
+  // 1. Check if the cut can be lowered to a LUT.
+
+  // Check if the cut has a single output.
+  if (cutOp.getNumResults() != 1)
+    return cutOp.emitError("expected single output");
+
+  // Check if every type is i1.
+  for (auto operand : cutOp.getOperands()) {
+    if (operand.getType() != rewriter.getI1Type())
+      return cutOp.emitError("expected i1 type");
+  }
+
+  for (auto result : cutOp.getResults()) {
+    if (result.getType() != rewriter.getI1Type())
+      return cutOp.emitError("expected i1 type");
+  }
```

Comment:
I'm not sure you're supposed to emit errors inside a rewrite pattern :thinking:. The dialect conversion framework might call these multiple times, and if multiple rewrites apply to the same op, only one has to succeed. But in that case you wouldn't want to see errors printed from the other patterns that failed to apply.

I think you could just move the entire code here into a `getOperation().walk([&](CutOp cutOp){ ... });` call inside the pass itself.

Reply:
yeah thanks. These are stale verification which I originally implemented with OpConversionPattern. I'll move these to walk.